### PR TITLE
feat: verify claims in provenance match the certificate

### PR DESCRIPTION
**Pull request by laurentsimon**
I made a change to cache the TUF root during the execution. This should help with regression tests (see function `TrustedRootSingleton()`).
Additional e2e tests need to be created https://github.com/slsa-framework/slsa-verifier/issues/571

closes https://github.com/slsa-framework/slsa-verifier/issues/580
closes https://github.com/slsa-framework/slsa-verifier/issues/570

### CODE REVIEW

File: verifiers/internal/gha/builder.go [hunk 1]
```diff
@@ -175,39 +190,309 @@ func verifyTrustedBuilderRef(id *WorkflowIdentity, ref string) error {
...
+	runURI, err := getExtension(cert, "1.3.6.1.4.1.57264.1.21", true)
+	if err != nil {
+		return nil, err
+	}
+	runID := strings.TrimPrefix(runURI, fmt.Sprintf("https://github.com/%s/actions/runs/", sourceRepository))
+
+	// Subject path.
+	if !strings.HasPrefix(cert.URIs[0].Path, "/") {
+		return nil, fmt.Errorf("%w: %s", serrors.ErrorInvalidFormat, cert.URIs[0].Path)
+	}
```
Comment (ianlewis) [inline verifiers/internal/gha/builder.go:457]: nit: save the trimmed path to a variable close to where it was validated.
```suggestion
	// Subject path.
	if !strings.HasPrefix(cert.URIs[0].Path, "/") {
		return nil, fmt.Errorf("%w: %s", serrors.ErrorInvalidFormat, cert.URIs[0].Path)
	}
	// Remove the starting '/'
	workflowRef = cert.URIs[0].Path[1:]
```

File: verifiers/internal/gha/builder.go [hunk 2]
```diff
@@ -175,39 +190,309 @@ func verifyTrustedBuilderRef(id *WorkflowIdentity, ref string) error {
...
+
+	// 1.3.6.1.4.1.57264.1.2: DEPRECATED.
+	// https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md#1361415726412--github-workflow-BuildTrigger-deprecated
+	deprecatedBuildTrigger, err := getExtension(cert, "1.3.6.1.4.1.57264.1.2", false)
+	if err != nil {
+		return nil, err
+	}
+	// 1.3.6.1.4.1.57264.1.20 | Build Trigger
+	// https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md#13614157264120--build-trigger
+	buildTrigger, err := getExtension(cert, "1.3.6.1.4.1.57264.1.20", true)
```
Comment (ianlewis) [inline verifiers/internal/gha/builder.go:298]: nit: Could you just get the string of the `ObjectIdentifier`?

```suggestion
	buildTrigger, err := getExtension(cert, OIDBuildTrigger.String(), true)
```

**Comment by ianlewis** [comment]
I didn't take a really close look at the tests but they looked mostly ok to me.

File: verifiers/internal/gha/builder.go [hunk 3]
```diff
@@ -175,39 +190,309 @@ func verifyTrustedBuilderRef(id *WorkflowIdentity, ref string) error {
...
+//nolint:gocyclo
 func GetWorkflowInfoFromCertificate(cert *x509.Certificate) (*WorkflowIdentity, error) {
 	if len(cert.URIs) == 0 {
-		return nil, errors.New("missing URI information from certificate")
+		return nil, fmt.Errorf("%w: missing URI information from certificate", serrors.ErrorInvalidFormat)
+	}
+
+	// 1.3.6.1.4.1.57264.1.2: DEPRECATED.
+	// https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md#1361415726412--github-workflow-BuildTrigger-deprecated
+	deprecatedBuildTrigger, err := getExtension(cert, "1.3.6.1.4.1.57264.1.2", false)
```
[the remark this answers is not in the document]
Comment (laurentsimon) [reply verifiers/internal/gha/builder.go:292]: I created `getAndValidateEqualClaims(cert *x509.Certificate, deprecatedOid, oid string) (string, error)`.
Created https://github.com/slsa-framework/slsa-verifier/issues/580 to improve on the string vs OID code.

**Comment by laurentsimon** [comment]
All comments addressed. PTAL

**Comment by laurentsimon** [comment]
@asraa can you take a look?

**Review by ianlewis — APPROVED** [review]
Sorry forgot to approve.

**Comment by laurentsimon** [comment]
I completely messed up the unit tests. Will take a look later

**Comment by laurentsimon** [comment]
@ianlewis not sure what's wrong with the linter. I've tried to disable the deprecated method check but it's still failing:
```
Error: directive `//nolint:staticcheck we want to disable SA1019 only to use deprecated methods but there is a bug in golangci-lint.` should match `//nolint:staticcheck we want to disable SA1019 only to use deprecated methods but there is a bug in golangci-lint.[:<comma-separated-linters>] [// <explanation>]` (nolintlint)
```

Context: https://github.com/golangci/golangci-lint/issues/741#issuecomment-784171870
I tried disabling the linter in builder.go and builder_test.go.

**Comment by ianlewis** [comment]
> @ianlewis not sure what's wrong with the linter. I've tried to disable the deprecated method check but it's still failing:
> 
> ```
> Error: directive `//nolint:staticcheck we want to disable SA1019 only to use deprecated methods but there is a bug in golangci-lint.` should match `//nolint:staticcheck we want to disable SA1019 only to use deprecated methods but there is a bug in golangci-lint.[:<comma-separated-linters>] [// <explanation>]` (nolintlint)
> ```
> 
> Context: [golangci/golangci-lint#741 (comment)](https://github.com/golangci/golangci-lint/issues/741#issuecomment-784171870) I tried disabling the linter in builder.go and builder_test.go.

The `//nolint` declarations are fairly particular and can't have a leading space and need to have a `//` delimiter before the comment message. I updated the comments.

**Comment by laurentsimon** [comment]
> The `//nolint` declarations are fairly particular and can't have a leading space and need to have a `//` delimiter before the comment message. I updated the comments.

Thanks!

**Pull request merged**